docs/src/wallet_to_wallet_with_tariscript.md

```diff
@@ -0,0 +1,205 @@
+| Commitment    |  \\( C_c = k_c \cdot G + v_c \cdot H  \\)     |
+| Features      | Fc                                            |
+| Rangeproof    | \\( RP_c \\) for \\( \hat{C}_c \\)            |
+| Script Hash   | \\( \sigma_c = H( \alpha_c) \\)               |
+| Script offset public key | \\( K_{Oc} \\)                     |
+
+| Transaction Kernel |       |
+|--------------------|-------|
+| Public Excess      | \\( X_s + K_b \\)    |
+| Signature          | \\( s, R \\)         |
+| Fee                | f                    |
+| Metadata           | m                    |
+
+Alice will now send this finalized transaction back to Bob so he has a copy of it and the script and both parties can now submit this transaction to the blockchain.
+
+### Bob after transaction broadcast
+
+After the transaction has been broadcast to the mempool Bob will have to monitor the blockchain to detect what height the transaction is mined at. Bob will then
+record this Height in his database for use when spending the UTXO.
+
+## Changes to Transaction messages and store data in Rust
+In [RFC](https://rfc.tari.com/RFC-0201_TariScript.html) the changes to the TransactionInput, TransactionOutput and Transaction structs are detailed but there are
+two more structs that will need to be updated to support TariScript.
+
+This is the struct that is sent by the Sender to the Recipient and needs to contain 2 new fields
+```rust,ignore
+pub struct SingleRoundSenderData {
+    /// The transaction id for the recipient
+    pub tx_id: u64,
+    /// The amount, in µT, being sent to the recipient
+    pub amount: MicroTari,
+    /// The offset public excess for this transaction
+    pub public_excess: PublicKey,
+    /// The sender's public nonce
+    pub public_nonce: PublicKey,
+    /// The transaction metadata
+    pub metadata: TransactionMetadata,
+    /// Plain text message to receiver
+    pub message: String,
+    
+    // NEW FIELDS
+    
+    /// Hash of the receivers UTXO script, \sigma
+    pub script_hash: Vec<u8,
+    /// Public script offset key chosen for Recipient, K_o
+    pub public_script_offset_key: PublicKey,
+```
+
+And the receiver will need to keep track of their spendable UTXO's using the following updated struct
+
+```rust,ignore
+pub struct UnblindedOutput {
+    pub value: MicroTari,
+    pub spending_key: BlindingFactor,
+    pub features: OutputFeatures,
+    
+    // NEW FIELDS
+    
+    /// The serialised script
+    script: Vec<u8>,
+    /// The script input data, if any
+    input_data: Vec<u8>,
+    /// The block height that the UTXO was mined
+    height : u64,
+    /// Script private key, k_s
+    script_private_key: PrivateKey,
+}
+```
+**NOTE:** The height in the previous struct will need to be updated AFTER the transaction negotiation is complete once the Receiver has detected the UTXO has been mined.
+
+
+## Questions
+1. Does the receiver need to get the script itself earlier than finding it in the Finalized transaction?
```

Comment:
This depends on what you are doing and what the script is. 
In a normal Alice spends to Bob with Bob's script key in the script, I would argue that Bob chooses the script.  I would argue that in most cases the receiver would like to or should create the script.  Only in multi-party cases or one-sided transactions should the sender choose the script.

Reply:
That's what I thought. I would like to keep the process described in this document as generic as possible so it should work for any script. So that does mean we need to add in one extra round of communication where Bob sends script data to Alice. When implementing standard MW Tx we can assume the NO_OP script and skip that step.

Reply:
@mikethetike What is your take on how we should handle exchanging the script before hand. In certain use cases we can just assume the script (Like a standard transaction) but if we want to keep the wallet implementation generic regarding the script we will need to include this extra round of communication. 

When we get to the implementation should we use the assumed NO_OP script for standard Tx's and then implement a second negotiation protocol for generic scripts?

Reply:
Actually thinking about this more the receiver actually doesn't even need to know the details of Bob's script to construct any part of the transaction because Bob is the one that commits to it in the beta key of his Rangeproof. So I think we can do it this way:
1. Alice send the initial sending data including the UTXO script offset K_O
2. Bob replies with his UTXO and the script he is applying to his UTXO
3. Alice can confirm the script Bob gives her is correct by checking the Rangeproof and if she agrees with the nature of the script then she completes the Tx

I guess the only different case would be a fancy script that requires Alice to give Bob some info to construct it, I suppose that can be part of the initial message?

@SWvheerden ^

Reply:
I agree with this. All Bob requires to complete the UTXO is the public script offset key. Alice can give him this. 
But from a sender point, in 99% of cases do we care what the script is? We only care about the script if it's some fancy script where there is extra stuff in it and only then if there is a chance we can be the spender. But most scripts will only be of concern for the receiver.